client/scala/src/main/scala/ai/verta/blobs/dataset/PathBlob.scala

```diff
@@ -0,0 +1,124 @@
+package ai.verta.blobs.dataset
+
+import ai.verta.swagger._public.modeldb.versioning.model._
+
+import java.security.{MessageDigest, DigestInputStream}
+import java.io.{File, FileInputStream}
+
+import scala.collection.mutable.HashMap
+import scala.util.{Failure, Success, Try}
+
+/** Captures metadata about files
+ *  @param paths list of filepaths or directory paths
+ */
+ case class PathBlob(private val paths: List[String]) extends Dataset {
+  private val BufferSize = 8192
+
+  private val metadataList = paths.map(expanduser)
+  .flatMap((path: String) => processPath(new File(path)))
+  .map(metadata => metadata.path -> metadata)
+
+  protected var contents = HashMap(metadataList: _*)
+
+  override def equals(other: Any) = other match {
+    case other: PathBlob => contents.equals(other.contents)
+    case _ => false
+  }
+
+
+  /** Hash the file's content
+   *  From https://stackoverflow.com/questions/41642595/scala-file-hashing
+   *  @param path filepath
+   */
+  private def hash(file: File) = Try {
+     val buffer = new Array[Byte](BufferSize)
+     val messageDigest = MessageDigest.getInstance("MD5")
+
+     val dis = new DigestInputStream(
+       new FileInputStream(file),
+       messageDigest
+     )
+
+     try {
+       while (dis.read(buffer) != -1) {}
+     } finally {
+       dis.close()
+     }
+
+     // Convert to hexadecimal
+     messageDigest.digest.map("%02x".format(_)).mkString
+   }
+
+   /** Get the metadata of path.
+    *  @param file a file object, representing the path
+    *  @return a list of components of file under the path
+    */
+   private def processPath(file: File): List[FileMetadata] = {
+     var files = List(file) // stack
+     var ret: List[FileMetadata] = List()
+
+     // non-recursive DFS to prevent stack overflow
+     while (files.length > 0) {
+       var fileToProcess = files.head
+       files = files.drop(1)
+
+       if (fileToProcess.isDirectory()) {
+         files = fileToProcess.listFiles().toList ::: files
+       }
+       else {
+         ret = processFile(fileToProcess) :: ret
+       }
+     }
+
+     ret
+   }
+
+  /** Extract the metadata of the file
+   *  If the file has an invalid path, exception is thrown immediately, and program stops (if not caught outside)
```

Comment:
Don't throw. Return a try. Always.

Reply:
OK I'll propagate the failure up all the way to the processPath function then.